unixfs/io/sizeadj.go

```diff
@@ -0,0 +1,112 @@
+package io
+
+import (
+	"errors"
+	"io"
+)
+
+func newSizeAdjReadSeekCloser(base ReadSeekCloser, size uint64) (*sizeAdjReadSeekCloser,error) {
+	r := &sizeAdjReadSeekCloser{base: base, size: int64(size)}
+	if r.size < 0 {
+		return nil, errors.New("sizeAdjReadSeekCloser: size limited 2^63 − 1")
+	}
+	return r, nil
+}
+
+type sizeAdjReadSeekCloser struct {
+	base   ReadSeekCloser
+	size   int64
+	offset int64
+}
+
+// Read implements the Read method as defined by io.Reader
+func (r *sizeAdjReadSeekCloser) Read(p []byte) (int, error) {
+	if r.offset >= r.size { // EOF
+		return 0, io.EOF
+	}
+	if int64(len(p)) > r.size-r.offset { // truncate
+		newsize := r.size - r.offset
+		p = p[:newsize]
+	}
+	n, err := r.base.Read(p)
+	if err != io.EOF { // only pad when we get an EOF
+		r.offset += int64(n)
+		return n, err
+	}
+	for ; n < len(p) && r.offset+int64(n) < r.size; n++ { // pad
+		p[n] = 0
+	}
+	r.offset += int64(n)
+	return n, io.EOF
+}
+
+// Seek implements the Seek method as defined by io.Seeker
+func (r *sizeAdjReadSeekCloser) Seek(offset int64, whence int) (int64, error) {
+	newOffset := r.offset
+	switch whence {
+	case io.SeekStart:
+		newOffset = offset
+	case io.SeekCurrent:
+		newOffset += offset
+	case io.SeekEnd:
+		newOffset = r.size - offset
+	}
+	if newOffset < 0 {
+		return -1, errors.New("Seek will result in negative position")
+	}
+	// Its easier just to always use io.SeekStart rather than
+	// correctly adjust offset for io.SeekCurrent and io.SeekEnd.
+	_, err := r.base.Seek(newOffset, io.SeekStart)
```

Comment:
what happens if `r.size == 1000`, but `r.base.Size() == 500` and we try to call `r.Seek(700)`? 

It seems like this would error out, when it should succeed and subsequent reads should return up to 300 bytes of zeros. I guess i'm not sure what errors Seek returns when you try to seek past its bounds.

Reply:
Actually it won't error out, from the documentation (https://golang.org/pkg/io/#Seeker):
> Seeking to an offset before the start of the file is an error. __Seeking to any positive offset is legal__, but the behavior of subsequent I/O operations on the underlying object is implementation-dependent.